### PR TITLE
APNG encoding optimization

### PR DESCRIPTION
**Pull request by sturkmen72**
related #26840

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/imgcodecs/src/grfmt_png.cpp [hunk 1]
```diff
@@ -1416,6 +1416,8 @@ void PngEncoder::deflateRectFin(unsigned char* zbuf, uint32_t* zsize, int bpp, i
 
 bool PngEncoder::writeanimation(const Animation& animation, const std::vector<int>& params)
 {
+    if (animation.frames.size() < 2)
+        return write(animation.frames.back(), params);
```
Comment (sturkmen72) [inline modules/imgcodecs/src/grfmt_png.cpp:1420]: should i replace with for better readibility
```
if (animation.frames.size() == 1)
       return write(animation.frames[0], params);
```

**Comment by asmorkalov** [comment]
I propose to add simple performance test.

**Comment by vrabaud** [comment]
An animation with 1 frame is not the same as one image (though it will be displayed the same).

**Comment by sturkmen72** [comment]
> An animation with 1 frame is not the same as one image (though it will be displayed the same).

It should be like that for webp and avif. But when an animated png saved with only one frame, the only thing that changes actually is  being 10 times slower.

**Comment by sturkmen72** [comment]
anyway let me work on this PR a bit more.  creating a performans test and maybe rethinking on one frame apng

**Comment by sturkmen72** [comment]
it was Z_BEST_COMPRESSION used for writing ( even writing with IMWRITE_PNG_COMPRESSION param)
now Z_BEST_SPEED ( 1 ) is default and IMWRITE_PNG_COMPRESSION param works

File: modules/imgcodecs/src/grfmt_png.cpp [hunk 2]
```diff
@@ -1416,8 +1418,9 @@ void PngEncoder::deflateRectFin(unsigned char* zbuf, uint32_t* zsize, int bpp, i
 
 bool PngEncoder::writeanimation(const Animation& animation, const std::vector<int>& params)
 {
-    int compression_level = 6;
-    int compression_strategy = IMWRITE_PNG_STRATEGY_RLE; // Default strategy
+    if (animation.frames.size() == 1)
```
Comment (vrabaud) [inline modules/imgcodecs/src/grfmt_png.cpp:1421]: I still think this is an error: a 1-frame APNG is not the same as a PNG. The spec allows for 1 frame APNG: "1 is a valid value for a single-frame APNG.", cf https://wiki.mozilla.org/APNG_Specification

Comment (sturkmen72) [reply modules/imgcodecs/src/grfmt_png.cpp]: @vrabaud i removed this condition. but fyi 1 frame apng still saved without acTL and fcTL

Comment (vrabaud) [reply modules/imgcodecs/src/grfmt_png.cpp]: Well, that is a bug in itself. Sure, usually, when you ask for a 1-frame APNG, the user probably wants a PNG but OpenCV should not be the one to decide.

**Comment by asmorkalov** [comment]
@sturkmen72 Please rebase and fix conflict after you PR merge.

File: modules/imgcodecs/src/grfmt_png.cpp [hunk 3]
```diff
@@ -1436,8 +1435,11 @@ bool PngEncoder::writeanimation(const Animation& animation, const std::vector<in
         }
     }
 
+    compression_level = compression_level < 0 ? Z_BEST_SPEED : compression_level;
```
Comment (asmorkalov) [inline modules/imgcodecs/src/grfmt_png.cpp:1436]: Why not just set Z_BEST_SPEED in constructor and remove the line here?

Comment (sturkmen72) [reply modules/imgcodecs/src/grfmt_png.cpp:1436]: then https://github.com/opencv/opencv/blob/4.x/modules/imgcodecs/src/grfmt_png.cpp#L910-L920 will be meaningless without `compression_level = -1;` let me look later with a fresh mind

File: modules/imgcodecs/perf/perf_png.cpp [hunk 4]
```diff
@@ -58,6 +58,51 @@ PERF_TEST(PNG, encode)
...
+    TEST_CYCLE() imencode(".png", src, buf, { IMWRITE_PNG_FILTER, GetParam() });
+
+    std::cout << "Encoded buffer size: " << buf.size()
+        << " bytes, Compression ratio: " << std::fixed << std::setprecision(2)
+        << (static_cast<double>(buf.size()) / (src.total() * src.channels())) * 100.0 << "%" << std::endl;
+
+    SANITY_CHECK_NOTHING();
+}
+
+PERF_TEST_P(PNG, DISABLED_params_filter_max_compression, testing::ValuesIn(PNGFilterFlags))
```
Comment (asmorkalov) [inline modules/imgcodecs/perf/perf_png.cpp:89]: Why is it disabled?

File: modules/imgcodecs/include/opencv2/imgcodecs.hpp [hunk 5]
```diff
@@ -211,6 +212,18 @@ enum ImwritePNGFlags {
        IMWRITE_PNG_STRATEGY_FIXED        = 4  //!< Using this value prevents the use of dynamic Huffman codes, allowing for a simpler decoder for special applications.
      };
 
+//! Imwrite PNG specific values for IMWRITE_PNG_FILTER parameter key
+enum ImwritePNGFilterFlags {
+       IMWRITE_PNG_NO_FILTERS            = 0,   //!< Disables all filtering. No filters will be applied when saving the image.
+       IMWRITE_PNG_FILTER_NONE           = 8,   //!< Applies no filter to the PNG image (useful when you want to save the raw pixel data without any compression filter).
```
Comment (asmorkalov) [inline modules/imgcodecs/include/opencv2/imgcodecs.hpp:218]: What is difference between IMWRITE_PNG_NO_FILTERS and IMWRITE_PNG_FILTER_NONE? Why do we need both?

File: modules/imgcodecs/include/opencv2/imgcodecs.hpp [hunk 6]
```diff
@@ -96,6 +96,7 @@ enum ImwriteFlags {
        IMWRITE_PNG_COMPRESSION     = 16, //!< For PNG, it can be the compression level from 0 to 9. A higher value means a smaller size and longer compression time. If specified, strategy is changed to IMWRITE_PNG_STRATEGY_DEFAULT (Z_DEFAULT_STRATEGY). Default value is 1 (best speed setting).
        IMWRITE_PNG_STRATEGY        = 17, //!< One of cv::ImwritePNGFlags, default is IMWRITE_PNG_STRATEGY_RLE.
        IMWRITE_PNG_BILEVEL         = 18, //!< Binary level PNG, 0 or 1, default is 0.
+       IMWRITE_PNG_FILTER          = 19, //!< One of cv::ImwritePNGFilterFlags, default is IMWRITE_PNG_FILTER_SUB.
```
Comment (asmorkalov) [inline modules/imgcodecs/include/opencv2/imgcodecs.hpp:99]: Are there any changes against 4.11 (APNG support) and 4.10 (old style PNG)?

Comment (asmorkalov) [reply modules/imgcodecs/include/opencv2/imgcodecs.hpp:99]: It'll be great to add some accuracy test for the new option. You can, for example, compare file sizes or byte array sizes to ensure the compression applied properly.

Comment (sturkmen72) [reply modules/imgcodecs/include/opencv2/imgcodecs.hpp:99]: > It'll be great to add some accuracy test for the new option. You can, for example, compare file sizes or byte array sizes to ensure the compression applied properly.

as far as i tested file sizes are various according used zlib or zlib-ng  ( maybe also used versions creates different sizes)

Comment (asmorkalov) [reply modules/imgcodecs/include/opencv2/imgcodecs.hpp:99]: Let's compare fast/best compression to ensure best is smaller the fast, for example.

File: modules/imgcodecs/include/opencv2/imgcodecs.hpp (hunk 5 above)
Comment (sturkmen72) [reply modules/imgcodecs/include/opencv2/imgcodecs.hpp:218]: they produces different sized outputs

File: modules/imgcodecs/include/opencv2/imgcodecs.hpp (hunk 6 above)
Comment (sturkmen72) [reply modules/imgcodecs/include/opencv2/imgcodecs.hpp:99]: > Are there any changes against 4.11 (APNG support) and 4.10 (old style PNG)?

default values are same as 4.10

File: modules/imgcodecs/perf/perf_png.cpp (hunk 4 above)
Comment (sturkmen72) [reply modules/imgcodecs/perf/perf_png.cpp]: when a common user run it takes so long time ( --perf_force_samples  default value is 100 )

File: modules/imgcodecs/perf/perf_png.cpp [hunk 7]
```diff
@@ -58,6 +61,30 @@ PERF_TEST(PNG, encode)
     SANITY_CHECK_NOTHING();
 }
 
+PERF_TEST_P(PNG, params,
+    testing::Combine(
+        testing::Values(IMWRITE_PNG_STRATEGY_DEFAULT, IMWRITE_PNG_STRATEGY_FILTERED, IMWRITE_PNG_STRATEGY_HUFFMAN_ONLY, IMWRITE_PNG_STRATEGY_RLE, IMWRITE_PNG_STRATEGY_FIXED),
+        testing::Values(IMWRITE_PNG_NO_FILTERS, IMWRITE_PNG_FILTER_NONE, IMWRITE_PNG_FILTER_SUB, IMWRITE_PNG_FILTER_UP, IMWRITE_PNG_FILTER_AVG, IMWRITE_PNG_FILTER_PAETH, IMWRITE_PNG_FAST_FILTERS, IMWRITE_PNG_ALL_FILTERS),
+        testing::Values(1, 4, 9)))
```
Comment (sturkmen72) [inline modules/imgcodecs/perf/perf_png.cpp:68]: `testing::Values(1, 9)` maybe better. Edge values give enough information about output file sizes and compression time

**Comment by sturkmen72** [comment]
@asmorkalov i will try to write tests using [PngSuite - Image filtering / PNG-files](http://www.schaik.com/pngsuite/pngsuite_fil_png.html)

**Comment by asmorkalov** [comment]
Ok, please try to use existing files in opencv-extra and do not add more. We have some subset of the suite already.

File: modules/imgcodecs/include/opencv2/imgcodecs.hpp (hunk 5 above)
Comment (sturkmen72) [reply modules/imgcodecs/include/opencv2/imgcodecs.hpp]: IMWRITE_PNG_NO_FILTERS removed. with libpng IMWRITE_PNG_NO_FILTERS and IMWRITE_PNG_ALL_FILTERS gives same results. with libspng IMWRITE_PNG_NO_FILTERS and IMWRITE_PNG_FILTER_NONE gives same results.

**Comment by asmorkalov** [comment]
@sturkmen72 Please rebase and fix conflicts.

**Review by asmorkalov — APPROVED** [review]

**Comment by asmorkalov** [comment]
@vrabaud I'm ready to merge the patch. Do you have any remarks?

File: modules/imgcodecs/perf/perf_png.cpp [hunk 8]
```diff
@@ -58,6 +61,30 @@ PERF_TEST(PNG, encode)
     SANITY_CHECK_NOTHING();
 }
 
+PERF_TEST_P(PNG, params,
+    testing::Combine(
+        testing::Values(IMWRITE_PNG_STRATEGY_DEFAULT, IMWRITE_PNG_STRATEGY_FILTERED, IMWRITE_PNG_STRATEGY_HUFFMAN_ONLY, IMWRITE_PNG_STRATEGY_RLE, IMWRITE_PNG_STRATEGY_FIXED),
+        testing::Values(IMWRITE_PNG_FILTER_NONE, IMWRITE_PNG_FILTER_SUB, IMWRITE_PNG_FILTER_UP, IMWRITE_PNG_FILTER_AVG, IMWRITE_PNG_FILTER_PAETH, IMWRITE_PNG_FAST_FILTERS, IMWRITE_PNG_ALL_FILTERS),
+        testing::Values(1, 6, 9)))
```
Comment (sturkmen72) [inline modules/imgcodecs/perf/perf_png.cpp:68]: @asmorkalov to reduce execution time i think `testing::Values(1, 6)` will give enough information.
9 is about 5-6 times slower than 6 and gives only about %1 lower file size.

File: modules/imgcodecs/src/grfmt_spng.cpp [hunk 9]
```diff
@@ -526,6 +530,8 @@ bool SPngEncoder::write(const Mat &img, const std::vector<int> &params)
     int width = img.cols, height = img.rows;
     int depth = img.depth(), channels = img.channels();
     volatile bool result = false;
+    bool set_compression_level = false;
```
Comment (vrabaud) [inline modules/imgcodecs/src/grfmt_spng.cpp:533]: those two should not be outside `if (ctx)` below right? I guess they should be kept right after `ihdr.width = width;` no ?

File: modules/imgcodecs/src/grfmt_spng.hpp [hunk 10]
```diff
@@ -51,6 +51,12 @@ class SPngEncoder CV_FINAL : public BaseImageEncoder
 
 protected:
     static int writeDataToBuf(void *ctx, void *user, void *dst_src, size_t length);
+
+private:
+    int m_compression_level;
```
Comment (vrabaud) [inline modules/imgcodecs/src/grfmt_spng.hpp:56]: Those do not need to be members right? They are only used in one function

File: modules/imgcodecs/src/grfmt_png.cpp [hunk 11]
```diff
@@ -858,6 +862,8 @@ bool  PngEncoder::write( const Mat& img, const std::vector<int>& params )
     int depth = img.depth(), channels = img.channels();
     volatile bool result = false;
     AutoBuffer<uchar*> buffer;
+    bool set_compression_level = false;
```
Comment (vrabaud) [inline modules/imgcodecs/src/grfmt_png.cpp:865]: shouldn't that be in `if( setjmp( png_jmpbuf ( png_ptr ) ) == 0 )` below ?

File: modules/imgcodecs/src/grfmt_spng.hpp (hunk 10 above)
Comment (sturkmen72) [reply modules/imgcodecs/src/grfmt_spng.hpp:56]: used in PngEncoder::write() and PngEncoder::writeanimation()

**Comment by asmorkalov** [comment]
@vrabaud Please take a look again.
